File: app/src/main/java/mx/dev/franco/automusictagfixer/utilities/SnackbarMessage.java
package mx.dev.franco.automusictagfixer.utilities;

import android.content.Context;
import androidx.annotation.StringRes;
import com.google.android.material.snackbar.Snackbar;
import java.lang.ref.WeakReference;
import mx.dev.franco.automusictagfixer.common.Action;

/**
 * A class that wraps a message or its id resource.
 */
public class SnackbarMessage<D> {
  private final String title;
  private final String body;
  private final String mainActionText;
  private final Action mainAction;
  private final D data;
  private final int duration;
  private final boolean dismissible;

  public SnackbarMessage(Builder<D> builder) {
    this.title = builder.title;
    this.body = builder.body;
    this.mainActionText = builder.mainActionText;
    this.data = builder.data;
    this.mainAction = builder.mainAction;
    this.duration = builder.duration;
    this.dismissible = builder.dismissible;
  }

  public String getTitle() {
    return title;
  }

  public String getBody() {
    return body;
  }

  public String getMainActionText() {
    return mainActionText;
  }

  public Action getMainAction() {
    return mainAction;
  }

  public boolean isDismissible() {
    return dismissible;
  }

  public int getDuration() {
    return duration;
  }

  public D getData() {
    return data;
  }

  public static class Builder<D> {
    private String title;
    private String body;
    private String mainActionText;
    private Action mainAction = Action.NONE;
    private D data;
    private int duration = Snackbar.LENGTH_SHORT;
    private boolean dismissible = true;
    private WeakReference<Context> contextRef;
    public Builder(Context context) {
      contextRef = new WeakReference<>(context);
    }

    public Builder<D> title(@StringRes int title) {
      this.title = contextRef.get().getString(title);
      return this;
    }

    public Builder<D> title(String title) {
      this.title = title;
      return this;
    }

    public Builder<D> body(@StringRes int body) {
      this.body = contextRef.get().getString(body);
      return this;
    }

    public Builder<D> body(String body) {
      this.body = body;
      return this;
    }

    public Builder<D> mainActionText(@StringRes int mainActionText) {
      this.mainActionText = contextRef.get().getString(mainActionText);
      return this;
    }

    public Builder<D> mainActionText(String mainActionText) {
      this.mainActionText = mainActionText;
      return this;
    }

    public Builder<D> data(D data) {
      this.data = data;
      return this;
    }

    public Builder<D> action(Action action) {
      mainAction = action;
      return this;
    }

    public Builder<D> duration(int duration) {
      this.duration = duration;
      return this;
    }

    public Builder<D> dismissible(boolean dismissible) {
      this.dismissible = dismissible;
      return this;
    }

    public SnackbarMessage<D> build() {
      SnackbarMessage snackbarMessage = new SnackbarMessage<>(this);
      contextRef.clear();
      contextRef = null;
      return snackbarMessage;
    }

  }
}

File: app/src/main/java/mx/dev/franco/automusictagfixer/utilities/SnackbarMessage.kt
package mx.dev.franco.automusictagfixer.utilities

import android.content.Context
import androidx.annotation.StringRes
import com.google.android.material.snackbar.Snackbar
import mx.dev.franco.automusictagfixer.common.Action
import java.lang.ref.WeakReference

/**
 * A class that wraps a message or its id resource.
 */
class SnackbarMessage<D>(builder: Builder<D>) {
    val title: String?
    val body: String?
    val mainActionText: String?
    val mainAction: Action
    val data: D
    val duration: Int
    val isDismissible: Boolean

    class Builder<D>(context: Context) {
        var title: String? = null
        var body: String? = null
        var mainActionText: String? = null
        var mainAction = Action.NONE
        var data: D? = null
        var duration = Snackbar.LENGTH_SHORT
        var dismissible = true
        private var contextRef: WeakReference<Context>?
        fun title(@StringRes title: Int): Builder<D> {
            this.title = contextRef!!.get()!!.getString(title)
            return this
        }

        fun title(title: String?): Builder<D> {
            this.title = title
            return this
        }

        fun body(@StringRes body: Int): Builder<D> {
            this.body = contextRef!!.get()!!.getString(body)
            return this
        }

        fun body(body: String?): Builder<D> {
            this.body = body
            return this
        }

        fun mainActionText(@StringRes mainActionText: Int): Builder<D> {
            this.mainActionText = contextRef!!.get()!!.getString(mainActionText)
            return this
        }

        fun mainActionText(mainActionText: String?): Builder<D> {
            this.mainActionText = mainActionText
            return this
        }

        fun data(data: D): Builder<D> {
            this.data = data
            return this
        }

        fun action(action: Action): Builder<D> {
            mainAction = action
            return this
        }

        fun duration(duration: Int): Builder<D> {
            this.duration = duration
            return this
        }

        fun dismissible(dismissible: Boolean): Builder<D> {
            this.dismissible = dismissible
            return this
        }

        fun build(): SnackbarMessage<D> {
            val snackbarMessage: SnackbarMessage<D> = SnackbarMessage(this)
            contextRef!!.clear()
            contextRef = null
            return snackbarMessage
        }

        init {
            contextRef = WeakReference(context)
        }
    }

    init {
        title = builder.title
        body = builder.body
        mainActionText = builder.mainActionText
        data = builder.data!!
        mainAction = builder.mainAction
        duration = builder.duration
        isDismissible = builder.dismissible
    }
}